modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -109,11 +109,30 @@ public void write(final ChannelHandlerContext ctx, final Object msg, final Chann
         assert msg instanceof Netty4HttpResponse : "Invalid message type: " + msg.getClass();
         boolean success = false;
         try {
-            List<Tuple<Netty4HttpResponse, ChannelPromise>> readyResponses = write((Netty4HttpResponse) msg, promise);
-            for (Tuple<Netty4HttpResponse, ChannelPromise> readyResponse : readyResponses) {
-                doWrite(ctx, readyResponse.v1(), readyResponse.v2());
+            final Netty4HttpResponse response = (Netty4HttpResponse) msg;
+            if (response.getSequence() != writeSequence) {
+                if (outboundHoldingQueue.size() >= maxEventsHeld) {
+                    int eventCount = outboundHoldingQueue.size() + 1;
+                    throw new IllegalStateException(
+                        "Too many pipelined events [" + eventCount + "]. Max events allowed [" + maxEventsHeld + "]."
+                    );
+                }
+                // response is not at the current sequence number so we add it to the outbound queue and return
+                outboundHoldingQueue.add(new Tuple<>(response, promise));
+                success = true;
+                return;
             }
+
+            // response is at the current sequence number and does not need to wait for any other response to be written so we write
+            // it out directly
+            doWrite(ctx, response, promise);
             success = true;
```

Comment:
Should this be set after the loop below?

Reply:
No in fact I consciously moved it here. Once we resolve the promise in `doWrite` (which we will do one way or another) we must not touch it again, that would always be a bug.

Reply:
Can we add a test that provokes that failure and then see that it is fixed by this PR? Seems we do not have tests that verify that we do not double notify failures and we used to sometimes do so?

Reply:
I don't think we can test this without creating an unrealistic situation. We should never be throwing here unless we exceed the outbound queue capacity to begin with. I really just fixed this as a cleanup to make it more obviously correct, unless we have a bug somewhere this doesn't change anything.